proposals/0386-package-access-modifier.md

```diff
@@ -129,47 +130,61 @@ swiftc -module-name App -package-name appPkg ...
-
-### Package Symbols and `@inlinable` Functions
+The Swift Package Manager already has a concept of a package identity string for every package.  This string is verified to be unique, and it already works as a package name, so SwiftPM will pass it down automatically.  Other build systems such as Bazel may need to introduce a new build setting for a package name.  Since it needs to be unique, a reverse-DNS name may be used to avoid clashing.
 
-`package` functions can be made `@inlinable`.  Just like with `@inlinable public`, not all symbols are usable within the `@inlinable package` function: they must be `open`, `public`, `package`, `@usableFromInline`, or `usableFromPackageInline`.
-
-`@usableFromPackageInline` is a new attribute which allows a symbol to be used from `@inlinable package` functions (that are defined in the same module) without having to make the symbol `package` or `public`.  It can be used anywhere that `@usableFromInline` can be used, but the attributes cannot be combined.  A `@usableFromPackageInline` symbol must be `internal`. For example: 
+If a target needs to be excluded from the package boundary or needes to be part of a subgroup within the package boundary, it can be set in a target setting, with a new parameter `group` in the manifest, like so: 
 
 ```
-@usableFromPackageInline func internalFuncA() {}
+  .target(name: "Game", dependencies: ["Engine"], group: .excluded)
+```
 
-@inlinable package func pkgUse() {
-    internalFuncA() // OK
-}
+The `group` setting is mapped to the following type:
 
-@inlinable public func publicUse() {
-    internalFuncA() // OK
+```
+enum Group {
+  case package
+  case excluded
+  case named(String)
 }
 ```
 
-The existing `@usableFromInline` attribute can be applied to `package` symbols as well as `internal` symbols.  This allows those symbols to be used from `@inlinable public` functions (that are defined anywhere in the same package) without having to make them `public`.  The Swift frontend will include `@usableFromInline package` symbols in the ordinary `.swiftinterface` for a module, not its `.package.swiftinterface`.
+The default value is `.package`, and the target is built with `-package-name [package_id]`.  If set to `.excluded`, no `-package-name` is passed when building the target, thus the target has no access to any package symbols; it essentially acts as if it's a client outside of the package. This would be useful for an example app or a black-box testing target in the same package.  If set to `.named([group_id])`, `-package-name [package_id].[group_id]` is passed, and the target can only access package symbols from other targets within the same group in the package.  This essentially allows "sub-packages" within the package.
```

Comment:
I would recommend not using `[...]` as the placeholder syntax here, because in the example `.named([group_id])`, it reads like a Swift array literal. Perhaps just using all-caps (`.named(GROUP_ID)` and `-package-name PACKAGE_ID.GROUP_ID` would be clearer?

Reply:
Do we have any concrete use cases for `.named()` like we do for `.excluded`?

I was trying to find some on the review thread, but all I could come up with is many mentions of a package that consists of multiple independent set of targets in a single package (some called them nuclei), but independent sets of targets shouldn't import modules from each other so they're already not seeing `public` symbols. Are we trying to design a workaround for the issue that the build system isn't separating modules based on the dependency declarations in the package manifest?

Reply:
The `.named(group_id)` can provide grouping within a package, where only modules within the same subgroup can see each other's package symbols. I agree that this grouping is not related to the original problem this proposal tries to solve, but might be a nice thing to add for better organization in a package.

Reply:
Right, I think I have two issues with that:
- for people who don't know which option to pick, it currently seems unclear what this option really does where the other two options are clearly defined, with example use cases
- for people who did want this option, I'm not sure if it really provides what they want, since public API remains accessible from targets outside of the group (which makes sense in the context of this proposal of course, but not so much in the larger context)

I wonder if we can just have `.package` and `.excluded` in this proposal and push `.named()` to future directions which could promise a potentially more holistic feature around organizing your packages into multiple subcomponents.

Reply:
I was just thinking the same thing :) -- introduce `.named()` in the future. We can see if `.excluded` is enough for v1 and address grouping later if needed.